### PR TITLE
Enable batch fetching in advance

### PR DESCRIPTION
**Pull request by jarandaf**

### CODE REVIEW

File: petastorm/pytorch.py [hunk 1]
```diff
@@ -118,7 +147,10 @@ def __iter__(self):
         self._in_iter = True
 
         try:
-            for batch in self._iter_impl():
+            iterator = self._iter_impl()
+            if self._max_prefetch > 1:
+                iterator = BackgroundIterator(iterator, prefetch=self.max_prefetch)
```
Comment (selitvin) [inline petastorm/pytorch.py:152]: Will the `BackgroundIterator` be properly destroyed and the thread joined when we exit the function (either nominal exit or with an exception?).

File: petastorm/pytorch.py [hunk 2]
```diff
@@ -100,11 +102,38 @@ def decimal_friendly_collate(batch):
 loader."
 
 
+class BackgroundIterator(threading.Thread):
+    """Prefetch iterator results. A thread iterates the original iterator and
+    populates a queue. Iterating over this background iterator just consumes the underlying
+    queue until no other result is available."""
+    def __init__(self, iterator, prefetch=1000):
```
Comment (selitvin) [inline petastorm/pytorch.py:109]: Is it really prefetching? Setting queue size does not guarantee that we will prefetch the data before user starts consuming it, does it? Perhaps we call the argument 'queue_size'? 
Frankly, I am not sure how prefetching helps steady-state throughput. Wouldn't it just eliminate some hiccups when the training starts at the expense of training starting a bit later? Isn't steady state throughput the only important characteristic here?

Comment (selitvin) [inline petastorm/pytorch.py:109]: Having default value of 1000 batches for the queue size maybe a bit too much, given a batch is a row-group, and a rowgroup of couple of hundreds MBs are common.

**Review by selitvin — COMMENTED** [review]
Looks like we are in the right direction! Left some questions.

File: petastorm/pytorch.py (hunk 2 above)
Comment (jarandaf) [reply petastorm/pytorch.py:109]: Loader's `_iter_impl` yields batches and not row groups, right? This is what is enqueued. It is true that depending on the queue size more or less row groups will be processed, but I expect this to be controlled via the queue size and the batch size.

Comment (jarandaf) [reply petastorm/pytorch.py:109]: You are right, probably prefetching is not the right word. As discussed in #740, the main motivation of this PR is to enable parallel batch building while training a model (otherwise the model will always have to wait for a batch to be available and this may take some time, specially if the dataset has a big number of columns). I have observed a ~3x speedup in data throughput with this change.

File: petastorm/pytorch.py (hunk 1 above)
Comment (jarandaf) [reply petastorm/pytorch.py:152]: Need to check this, thank you for the heads up.

File: petastorm/pytorch.py (hunk 2 above)
Comment (selitvin) [reply petastorm/pytorch.py:109]: Sure - I see how this can speed up the training. This is a good change. 

In my understanding we are not really doing prefetching here (depending on the timing, the consumer might try to fetch the first batch before the thread has populated it, i.e. nothing was prefetched). 

If you are ok with just changing the name from prefetching to queue size, everything will fall in place then.

Comment (selitvin) [reply petastorm/pytorch.py:109]: You are absolutely right.

File: petastorm/pytorch.py (hunk 1 above)
Comment (selitvin) [reply petastorm/pytorch.py:152]: I know it could be tricky, but would be good to also test these aspects in a unit test or two.

**Comment by jarandaf** [comment]
@selitvin could you please have a look? Thank you!

File: petastorm/pytorch.py [hunk 3]
```diff
@@ -100,11 +102,42 @@ def decimal_friendly_collate(batch):
...
+        self.name = "background_iterator"
+        self.queue = Queue(queue_size)
+        self.iterator = iterator
+        self.stop = threading.Event()
+        self.start()
+
+    def run(self):
+        while not self.stop.isSet():
+            for item in self.iterator:
+                self.queue.put(item)
```
Comment (selitvin) [inline petastorm/pytorch.py:120]: I don't think we can use blocking puts and gets and end up with a solution that is robust to deadlocks. Let's see if this works:
- Producer ended up filling the queue and waits on a blocking put.
- Consumer fails. Calls `iterator.stop.set()`, however since .put is blocking within an iterator queue, the event is never checked and the thread is not shut down.

Another scenario: 
- The queue is empty, hence consumer waits on a blocking `.get`. 
- However, producer raises an exception. The thread dies and the consumer is stuck forever on a .get.

I think a robust implementation for a `BackgroundIterator` could get pretty tricky. All these edge cases need to be carefully tested as these kind of failures would be hard to catch in production.

File: petastorm/pytorch.py [hunk 4]
```diff
@@ -117,15 +150,20 @@ def __iter__(self):
...
+            for batch in iterator:
                 yield batch
         except Exception as e:
             self._error = e
             logger.error('Iteration on Petastorm DataLoader raise error: %s', repr(e))
             raise
         finally:
             self._in_iter = False
+            if isinstance(iterator, BackgroundIterator):
+                iterator.stop.set()
```
Comment (selitvin) [inline petastorm/pytorch.py:166]: Let's make `stop` a private member (i.e. `_stop`) and add an API to the BackgroundIterator that performs stop (encapsulation principle).

File: petastorm/pytorch.py [hunk 5]
```diff
@@ -100,11 +102,42 @@ def decimal_friendly_collate(batch):
...
 
 
+class BackgroundIterator(threading.Thread):
+    """Prefetch iterator results. A thread iterates the original iterator and
+    populates a queue. Iterating over this background iterator just consumes the underlying
+    queue until no other result is available."""
+    def __init__(self, iterator, queue_size=1000):
+        threading.Thread.__init__(self)
+        self.name = "background_iterator"
+        self.queue = Queue(queue_size)
```
Comment (selitvin) [inline petastorm/pytorch.py:112]: Let's mark all data members that are not intended to be exposed to BackgroundIterator users as private (`_` prefix).

**Review by selitvin — CHANGES_REQUESTED** [review]

**Comment by chongxiaoc** [comment]
Just noticed this nice work. @jarandaf Thanks!
Is it doable to support shuffle to be parallel as well?
I think usually shuffling is the bottleneck and petastorm uses single thread for that.
